lib/CL/devices/almaif/almaif.cc

```diff
@@ -322,74 +316,49 @@ cl_int pocl_almaif_init(unsigned j, cl_device_id dev, const char *parameters) {
   D->BaseAddress = strtoull(paramToken, NULL, 0);
 
   std::string supportedList;
-  char *device_init_file = NULL;
+  std::string device_init_file = "";
   if (D->BaseAddress != 0xE) {
     paramToken = strtok_r(NULL, ",", &savePtr);
     assert(paramToken);
-    device_init_file = (char *)malloc(strlen(paramToken) + 1);
-    assert(device_init_file);
-    strcpy(device_init_file, paramToken);
-    POCL_MSG_PRINT_ALMAIF("Enabling device with device init file name %s",
-                          device_init_file);
+    device_init_file = paramToken;
+    POCL_MSG_PRINT_ALMAIF("Enabling device with device init file name %s\n",
+                          device_init_file.c_str());
   }
 
   bool enable_compilation = false;
 
-  while ((paramToken = strtok_r(NULL, ",", &savePtr))) {
-    auto token = strtoul(paramToken, NULL, 0);
-    BuiltinKernelId kernelId = static_cast<BuiltinKernelId>(token);
-
-    bool found = false;
-    for (size_t i = 0; i < BIKERNELS; ++i) {
-      if (pocl_BIDescriptors[i].KernelId == kernelId) {
-        if (supportedList.size() > 0)
-          supportedList += ";";
-        supportedList += pocl_BIDescriptors[i].name;
-        D->SupportedKernels.insert(&pocl_BIDescriptors[i]);
-        found = true;
-        break;
-      }
-    }
-    if (kernelId == POCL_CDBI_JIT_COMPILER) {
-      enable_compilation = true;
-    } else if (!found) {
-      POCL_ABORT("almaif: Unknown Kernel ID (%lu) given\n", token);
-    }
-  }
-  free(scanParams);
-
   // almaif devices are little endian by default, but the emulation device is
   // host dependant
   dev->endian_little = D->BaseAddress == 0xE ? !(WORDS_BIGENDIAN) : CL_TRUE;
   if (D->BaseAddress == 0xE) {
     dev->long_name = (char *)"almaif emulation device";
   }
 
-  dev->builtin_kernel_list = strdup(supportedList.c_str());
-  dev->num_builtin_kernels = D->SupportedKernels.size();
-  pocl_setup_builtin_kernels_with_version(dev);
 
   if (!pocl_offline_compile) {
 
-    POCL_MSG_PRINT_ALMAIF(
-        "almaif: accelerator at 0x%zx with %zu builtin kernels (%s)\n",
-        D->BaseAddress, D->SupportedKernels.size(), dev->builtin_kernel_list);
     // Recognize whether we are emulating or not
     if (D->BaseAddress == EMULATING_ADDRESS) {
       D->Dev = new EmulationDevice();
-    }
+    } else if (D->BaseAddress == 0xA) {
```

Comment:
What is this magic number?

Reply:
This is for commandline parameters. It just so I can rapidly switch between different AlmaIFDevice backends. I have some magic numbers for that. So it works something like this:
POCL_ALMAIF0_PARAMETERS=0xA --> XrtDevice
POCL_ALMAIF0_PARAMETERS=0xF --> DBDevice
POCL_ALMAIF0_PARAMETERS=0xE --> EmulationDevice
POCL_ALMAIF0_PARAMETERS=0xB --> TTASimDevice

Reply:
Best use an enum/macro to make this clearer. Magic numbers are baaad.

Reply:
Converted to enum.
This is a little bit hacky way to switch between devices. But maybe it's fine, if all the AlmaIFDevice types become accessible via the AFOCL database's JSON-file. So maybe in the future, the JSON-file will contain all this metadata, so this would no longer be user-facing